### PR TITLE
System: Add NewControllerSystem

### PR DESCRIPTION
**Pull request by ao-david**
NewControllerSystem returns a controller system based on the passed inputs and functional options. The ControllerSystem is used to link together a controller device, its variables, sensors, actuators, and eventually connected peripherals.

See also: #207 #209 #210

### CODE REVIEW

File: system/controllers.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	for i, opt := range options {
+		err := opt(sys)
+		if err != nil {
+			return nil, fmt.Errorf("failed to apply option[%d]: %w", i, err)
+		}
+	}
+
+	// Append site/device information to actuators, sensors, and variables.
+	for _, variable := range sys.Variables {
+		log.Println("updating vars.")
```
Comment (scruzin) [inline system/controllers.go:51]: Did you mean to keep this log statement?

File: system/controllers.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+		variable.Skey = skey
+		variable.Scope = strings.ReplaceAll(sys.Controller.MAC(), ":", "")
+	}
+	for _, sensor := range sys.Sensors {
+		sensor.Mac = sys.Controller.Mac
```
Comment (scruzin) [inline system/controllers.go:56]: The sensor must have the MAC set before PutSensor is called. Is that the case?

File: system/controllers.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	}
+	for _, actuator := range sys.Actuators {
+		actuator.Mac = sys.Controller.Mac
```
Comment (scruzin) [inline system/controllers.go:59]: Ditto for actuators.

File: system/controllers.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	Variables   []*model.Variable
+	Sensors     []*model.SensorV2
+	Actuators   []*model.ActuatorV2
+	Peripherals []*model.Device
+}
+
+// Option represents functional options that can be passed to NewControllerSystem.
+type Option func(*ControllerSystem) error
+
+func NewControllerSystem(skey int64, MAC, name string, options ...Option) (*ControllerSystem, error) {
```
Comment (scruzin) [inline system/controllers.go:24]: Function comment, e.g.,
`// NewControllerSystem creates a controller device with optional variables, sensors, actuators, etc.`

File: system/controllers.go (hunk 2 above)
Comment (ao-david) [reply system/controllers.go:56]: In the current implementation it is up to the user to put the controller system (and all components) into the datastore, so currently, yes this happens before then.

File: system/controllers.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+//
+// MAC and name refer to the MAC Address and name of the Controller which will be the heart of
+// the ControllerSystem.
+func NewControllerSystem(skey int64, MAC, name string, options ...Option) (*ControllerSystem, error) {
+	// Check MAC.
+	if model.MacEncode(MAC) == 0 {
+		return nil, model.ErrInvalidMACAddress
+	}
+
+	// Create a ControllerSystem.
```
Comment (saxon-milton) [inline system/controllers.go:34]: Redundant comment.

File: system/controllers.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+// Option represents functional options that can be passed to NewControllerSystem.
+type Option func(*ControllerSystem) error
+
+// NewControllerSystem returns a new ControllerSystem with the given options. It is the callers
+// responsibility to put the components into the datastore.
+//
+// MAC and name refer to the MAC Address and name of the Controller which will be the heart of
+// the ControllerSystem.
+func NewControllerSystem(skey int64, MAC, name string, options ...Option) (*ControllerSystem, error) {
+	// Check MAC.
```
Comment (saxon-milton) [inline system/controllers.go:29]: Redundant comment.

**Review by saxon-milton — CHANGES_REQUESTED** [review]

File: system/controllers.go (hunk 2 above)
Comment (ao-david) [reply system/controllers.go:59]: This functionality will be in a future change I think. I agree that having all the components be abstracted as one system for datastore operations is certainly the way to go. At the moment though, I am just creating the system.

File: system/controllers_test.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+		{
+			name:           "Invalid MAC without options",
+			skey:           testSiteKey,
+			mac:            "00:11:22:33:44",
+			ControllerName: "TestController",
+			options:        nil,
+			wantErr:        true,
+			expectedSystem: nil,
+		},
+	}
```
[the remark this answers is not in the document]
Comment (ao-david) [reply system/controllers_test.go:49]: I will add options in another change, then add test cases then

**Comment by ao-david** [comment]
@saxon-milton The changes to follow this will be adding functional options, at which point these functions will get further testing with realistic options being passed.

**Review by saxon-milton — APPROVED** [review]

**Pull request merged**